### PR TITLE
feat(interactive): pattern builder

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR adds in "Pattern Builder Mode" to Interactive Mode, which allows you to customize the Semgrep patterns that you "build" in the Interactive REPL.

## Why:
I figured that it would feel bad for people who make searches and want to later refine them, or rewrite parts of it, who don't want to have to restart the entire process and retype things. Essentially, it's just to give you more power and customizability, and enhance your user experience.

## How:
We have a notion of two primary modes, "Navigator Mode" and "Pattern Builder Mode".

In Navigator Mode, the up, down, left, and right arrow keys are used for traversing the files and matches. Entering a pattern and pressing enter will cause a new pattern to be added (depending on `All` or `Any` mode) to the pattern tree.

In Pattern Builder Mode, the up and down arrow keys are instead used to traverse the previously-built pattern tree. There,
you can edit the parts of the pattern tree in place, in case you want to modify your query.

We achieved this by adding a distinction between the `on_event_navigator` and `on_event_pattern`, in which case alternate code paths are taken. Logic was added to properly highlight between the Pattern Builder and file menus, to load in the currently-edited pattern string to the interactive REPL, to switch modes smoothly, and to traverse the pattern tree at all.

The last one required a more sophisticated, recursive pattern zipper. It's a zipper of zippers, essentially. Currently, you are able to traverse to the actual `any` and `all` nodes as well, because I have some future ideas for what you might be able to do with that.

## Test plan:
https://asciinema.org/a/MYnltx59TAQFQEVBOIlj8rwFY

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [X] Tests included or PR comment includes a reproducible test plan
- [X] Documentation is up-to-date
- [ ] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: src/osemgrep/cli_interactive/Framed_zipper.mli [hunk 1]
```diff
@@ -52,3 +57,5 @@ val is_empty : 'a t -> bool
 val length : 'a t -> int
 val absolute_position : 'a t -> int
 val relative_position : 'a t -> int
+val is_leftmost : 'a t -> bool
```
Comment (aryx) [inline src/osemgrep/cli_interactive/Framed_zipper.mli]: are you talking about the pointer or the frame? the pointer is at the beginning of the list, or the frame (whatever its size) is at the start?

File: src/osemgrep/cli_interactive/Interactive_subcommand.ml [hunk 2]
```diff
@@ -36,12 +36,32 @@ type command =
   (* actual commands *)
   | Exit
 
-(* interactive formula *)
-type iformula =
+(* interactive formula, but in a zipper!
+
+   When in Pattern Builder mode, we want to be able to traverse the
+   formula, up and down. This means we need another zipper. We can
+   use `Framed_zipper` of frame 1 to mimic a true zipper, again.
```
Comment (aryx) [inline src/osemgrep/cli_interactive/Interactive_subcommand.ml]: Maybe define a Zipper.ml where you simply reuse Framed_zipper.ml but set the frame to 1?

**Review by aryx — APPROVED** [review]

File: src/osemgrep/cli_interactive/Framed_zipper.mli (hunk 1 above)
Comment (brandonspark) [reply src/osemgrep/cli_interactive/Framed_zipper.mli]: left a comment explaining

File: src/osemgrep/cli_interactive/Unit_framed_zipper.ml [hunk 3]
```diff
@@ -148,3 +148,22 @@ let%expect_test "position" =
   [%expect {|
     [3 | 3, 4, (5)]
     absolute: 4, relative: 2 |}]
+
+let%expect_test "top_bottom" =
+  let str_of_bool b = if b then "T" else "F" in
+  let zipper = FZ.of_list 3 [ 1; 2; 3 ] in
+  Common.pr2 (str_of_bool (FZ.is_top zipper));
```
Comment (aryx) [inline src/osemgrep/cli_interactive/Unit_framed_zipper.ml:155]: Honestly I prefer assert FZ.is_top_zipper than those pr2 str_of_bool stuff.
Not sure this is a good use case for ppx_expect tests.

**Pull request merged**